Provider writes logging to disk — in a Unix-specific way (doesn't work on Windows)

**What steps did you take:**
```
terraform {
  required_version = "~> 1.0"
  required_providers {
    carvel = {
      source  = "vmware-tanzu/carvel"
      version = "0.11.0"
    }
  }
}
```

```
terraform init
```
```
Initializing the backend...

Initializing provider plugins...
- Reusing previous version of vmware-tanzu/carvel from the dependency lock file
- Using previously-installed vmware-tanzu/carvel v0.11.0

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```
```
terraform providers schema -json
╷
│ Error: Failed to load plugin schemas
│
│ Error while loading schemas for plugin components: Failed to obtain provider schema: Could not load the schema for provider registry.terraform.io/vmware-tanzu/carvel: failed to instantiate provider
│ "registry.terraform.io/vmware-tanzu/carvel" to obtain schema: Unrecognized remote plugin message:
│
│ This usually means that the plugin is either invalid or simply
│ needs to be recompiled to support the latest protocol...
╵


Stack trace from the terraform-provider-carvel_v0.11.0.exe plugin:

panic: open /tmp/terraform-provider-carvel.log: The system cannot find the path specified.

goroutine 1 [running]:
github.com/vmware-tanzu/terraform-provider-carvel/pkg/logger.MustNewFileRoot(0xdc3c41, 0x22, 0x413c98)
        github.com/vmware-tanzu/terraform-provider-carvel/pkg/logger/root.go:42 +0x9f
github.com/vmware-tanzu/terraform-provider-carvel/pkg/provider.Provider(0xc0003d0c30, 0xc00008ff18)
        github.com/vmware-tanzu/terraform-provider-carvel/pkg/provider/provider.go:14 +0x41
github.com/hashicorp/terraform/plugin.Serve(0xc00008ff58)
        github.com/hashicorp/terraform@v0.14.0/plugin/serve.go:60 +0x1fa
main.main()
        github.com/vmware-tanzu/terraform-provider-carvel/cmd/main.go:9 +0x59

Error: The terraform-provider-carvel_v0.11.0.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

**What happened:**
Provider panicked.

**What did you expect:**
Provider to not panic.

**Anything else you would like to add:**

The simple fix here would likely be replacing the static `/tmp` with [`os.TempDir()`](https://pkg.go.dev/os#TempDir), [`os.CreateTemp()`](https://pkg.go.dev/os#CreateTemp) or similar.

However, the vast majority of provider do not create log files on its own (certainly not by default). If/when debugging is necessary, Terraform itself comes with tools which allow access to logs, e.g. using `TF_LOG_PROVIDER` or `TF_LOG` on older versions. This can be controlled by the end-user. See https://developer.hashicorp.com/terraform/plugin/log/managing for more.

AFAICT it's not currently documented as best practice (that providers don't re-implement logging on their own) but I _would_ say it is implied from the fact that Terraform provides its own logging interface and documents it. It is also implied that providers don't introduce side effects - i.e. don't do anything other than what users may reasonably expect on the target system. Meaning that if they're supposed to interact with a remote API, then there should be no OS-specific behaviours or even need for any writable filesystem.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -1,0 +1,21 @@
+---
+name: Tests
+
+on:
+  pull_request:
+    types: ['opened', 'reopened', 'synchronize']
+
+jobs:
+  go-test:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install Go
+        uses: actions/setup-go@v4
+        with:
+          go-version-file: go.mod
+      - name: Run tests
+        run: go test ./...

--- a/pkg/kapp/kapp.go
+++ b/pkg/kapp/kapp.go
@@ -80,7 +80,7 @@ func (t *Kapp) Diff() (string, string, error) {
 
 	if err == nil {
 		return "", stderrStr, fmt.Errorf("Executing kapp: Expected "+
-			"non-0 exit code (stderr: %s)", err, stderrStr)
+			"non-0 exit code (stderr: %s)", stderrStr)
 	}
 
 	if exitError, ok := err.(*goexec.ExitError); ok {

--- a/pkg/provider/provider.go
+++ b/pkg/provider/provider.go
@@ -1,6 +1,9 @@
 package provider
 
 import (
+	"os"
+	"path/filepath"
+
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/hashicorp/terraform/terraform"
 	"github.com/vmware-tanzu/terraform-provider-carvel/pkg/kapp"
@@ -11,7 +14,7 @@ import (
 )
 
 func Provider() terraform.ResourceProvider {
-	logger := logger.MustNewFileRoot("/tmp/terraform-provider-carvel.log")
+	logger := logger.MustNewFileRoot(filepath.Join(os.TempDir(), "terraform-provider-carvel.log"))
 	yttLogger := logger.WithLabel("ytt")
 	kbldLogger := logger.WithLabel("kbld")
 

